text/0000-fix-error.md

```diff
@@ -0,0 +1,246 @@
+- Feature Name: fix_error
+- Start Date: 2018-07-18
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Change the `std::error::Error` trait to improve its usability. Introduce a
+backtrace module to the standard library to provide a standard interface for
+backtraces.
+
+# Motivation
+[motivation]: #motivation
+
+The `Error` trait has long been known to be flawed in several respects. In
+particular:
+
+1. The required `description` method is limited, usually, to returning static
+strings, and has little utility that isn't adequately served by the required
+`Display` impl for the error type.
+2. The signature of the `cause` method does not allow the user to downcast the
+cause type, limiting the utility of that method unnecessarily.
+3. It provides no standard API for errors that contain backtraces (as some
+users' errors do) to expose their backtraces to end users.
+
+We propose to fix this by deprecating the existing methods of `Error` and adding
+two new, provided methods. As a result, the undeprecated portion of the `Error`
+trait would look like this:
+
+```rust
+trait Error: Display + Debug {
+    fn backtrace(&self) -> Option<&Backtrace> {
+        None
+    }
+
+    fn source(&self) -> Option<&dyn Error + 'static> {
+        None
+    }
+}
+```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## The new API of the Error trait
+
+The new API provides three main components:
+
+1. The Display and Debug impls, for printing error messages. Ideally, the
+   Display API would be focused on *end user* messages, whereas the Debug impl
+   would contain information relevant to the programmer.
+2. The backtrace method. If the Error contains a backtrace, it should be exposed
+   through this method. Errors are not required to contain a backtrace, and are
+   not expected to.
+3. The `source` method. This returns another Error type, which is the underlying
+   source of this error. If this error has no underlying source (that is, it is
+   the "root source" of the error), this method should return none.
+
+## The backtrace API
+
+This RFC adds a new `backtrace` module to std, with one type, with this API:
+
+```rust
+pub struct Backtrace {
+    // ...
+}
+
+impl Backtrace {
+    pub fn new() -> Backtrace {
+        // ...
+    }
+}
+
+impl Display for Backtrace {
+    // ...
+}
+
+impl Debug for Backtrace {
+    // ...
+}
+```
+
+This minimal initial API is just intended for printing backtraces for end users.
+In time, this may grow the ability to visit individual frames of the backtrace.
+
+### Backtrace controlling environment variables
+
+Today, the `RUST_BACKTRACE` controls backtraces generated by panics. After this
+RFC, it also controls backtraces generated in the standard library: no backtrace
+will be generated when calling `Backtrace::new` unless this variable is set.
```

Comment:
Does this mean `Backtrace::new` should return an Option, or that it will simply not be called by any code inside std (and always produces a backtrace when called)?